sql/src/main/java/io/crate/planner/UnionPlan.java

```diff
@@ -0,0 +1,170 @@
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Plan for Union which uses a MergePhase to combine the results of two plans (= two inputs).
+ */
+public class UnionPlan implements Plan, ResultDescription {
+
+    private final Plan left;
+    private final Plan right;
+
+    private final MergePhase mergePhase;
+
+    private int unfinishedLimit;
+    private int unfinishedOffset;
+    private int numOutputs;
+
+    private final int maxRowsPerNode;
+
+    @Nullable
+    private PositionalOrderBy orderBy;
+
+    /**
+     * Create a Union Plan
+     *
+     * @param unfinishedLimit the limit a parent must apply after a merge to get the correct result
+     * @param unfinishedOffset the offset a parent must apply after a merge to get the correct result
+     *
+     * If the data should be limited as part of the Merge, add a {@link io.crate.planner.projection.TopNProjection},
+     * if possible. If the limit of the TopNProjection is final, unfinishedLimit here should be set to NO_LIMIT (-1)
+     *
+     * See also: {@link ResultDescription}
+     *
+     */
+    public UnionPlan(Plan left,
+                     Plan right,
+                     MergePhase mergePhase,
+                     int unfinishedLimit,
+                     int unfinishedOffset,
+                     int numOutputs,
+                     int maxRowsPerNode) {
+        this.left = left;
+        this.right = right;
+        Preconditions.checkArgument(mergePhase.numInputs() == 2,
+            "Number of inputs of MergePhase needs to be two.");
+        this.mergePhase = mergePhase;
+        this.unfinishedLimit = unfinishedLimit;
+        this.unfinishedOffset = unfinishedOffset;
+        this.numOutputs = numOutputs;
+        this.maxRowsPerNode = maxRowsPerNode;
+    }
+
+    @Override
+    public <C, R> R accept(PlanVisitor<C, R> visitor, C context) {
+        return visitor.visitUnionPlan(this, context);
+    }
+
+    @Override
+    public UUID jobId() {
+        return mergePhase.jobId();
+    }
+
+    @Override
+    public void addProjection(Projection projection) {
+        mergePhase.addProjection(projection);
+        numOutputs = projection.outputs().size();
+    }
+
+    @Override
+    public void addProjection(Projection projection,
+                              int unfinishedLimit,
+                              int unfinishedOffset,
+                              @Nullable PositionalOrderBy unfinishedOrderBy) {
+        addProjection(projection);
+        this.unfinishedLimit = unfinishedLimit;
+        this.unfinishedOffset = unfinishedOffset;
+        this.orderBy = unfinishedOrderBy;
+    }
+
+    @Override
+    public ResultDescription resultDescription() {
+        return this;
+    }
+
+    @Override
+    public void setDistributionInfo(DistributionInfo distributionInfo) {
+        mergePhase.distributionInfo(distributionInfo);
+    }
+
+    public MergePhase mergePhase() {
```

Comment:
very Minor: I would move these 3 methods to the end to keep all the Overriden methods together